components/src/main/java/uk/gov/android/ui/components/m3/ButtonsM3.kt

```diff
@@ -0,0 +1,272 @@
+import uk.gov.android.ui.theme.m3.GdsTheme
+import uk.gov.android.ui.theme.minimumTouchTarget
+import uk.gov.android.ui.theme.smallPadding
+
+@Composable
+fun GdsButtonM3(
+    buttonParameters: ButtonParametersM3,
+    colors: ColorScheme = MaterialTheme.colorScheme
+) {
+    buttonParameters.apply {
+        key(this) {
+            val buttonColors = buttonType.buttonColour()
+            Button(
+                colors = buttonColors,
+                elevation = ButtonDefaults.buttonElevation(
+                    defaultElevation = 0.dp
+                ),
+                modifier = Modifier
+                    .semantics(mergeDescendants = true) {}
+                    .heightIn(min = minimumTouchTarget)
+                    .widthIn(min = minimumTouchTarget)
+                    .then(modifier),
+                onClick = onClick,
+                shape = RoundedCornerShape(0.dp),
+                enabled = enabled
+            ) {
+                buttonContent(buttonParameters, colors).invoke(this)
+            }
+        }
+    }
+}
+
+private fun buttonContent(
+    parameters: ButtonParametersM3,
+    colors: ColorScheme
+): @Composable RowScope.() -> Unit = {
+    if (parameters.buttonType is ButtonTypeM3.ICON &&
+        !parameters.buttonType.iconParameters.imagePositionAtEnd
+    ) {
+        DisplayIcon(
+            iconButtonType = parameters.buttonType,
+            colors = colors,
+            imagePositionAtEnd = false,
+            isPrimary = parameters.buttonType.buttonType is ButtonTypeM3.PRIMARY
+        )
+    }
+
+    Text(
+        fontWeight = parameters.buttonType.fontWeight,
+        style = parameters.textStyle ?: MaterialTheme.typography.labelMedium,
+        text = stringResource(id = parameters.text)
+    )
+
+    if (parameters.buttonType is ButtonTypeM3.ICON &&
+        parameters.buttonType.iconParameters.imagePositionAtEnd
+    ) {
+        DisplayIcon(
+            iconButtonType = parameters.buttonType,
+            colors = colors,
+            imagePositionAtEnd = true,
+            isPrimary = parameters.buttonType.buttonType is ButtonTypeM3.PRIMARY
+        )
+    }
+}
+
+@Composable
+private fun DisplayIcon(
+    iconButtonType: ButtonTypeM3.ICON,
+    colors: ColorScheme,
+    imagePositionAtEnd: Boolean,
+    isPrimary: Boolean
+) {
+    val modifier = if (imagePositionAtEnd) {
+        Modifier
+            .padding(start = smallPadding)
+    } else {
+        Modifier
+            .padding(end = smallPadding)
+    }
+    val fgColor = if (isPrimary) {
+        MaterialTheme.colorScheme.onPrimary
+    } else {
+        MaterialTheme.colorScheme.onSecondary
+    }
+    GdsIconM3(
+        parameters = iconButtonType.iconParameters.copy(
+            foreGroundColor = fgColor,
+            modifier = modifier.then(
+                iconButtonType.iconParameters.modifier
+            )
+        ),
+        colors = colors
+    )
+}
+
+@Keep
+data class ButtonParametersM3(
```

Comment:
Any reason not to make this use the `.generate()` approach here?

Reply:
Yes. GdsButtonM3 (and GdsIconM3) takes the composable MaterialTheme.colorScheme as a parameter. If you attempt to add this to the  ButtonParametersM3 constructor parameter you will get a compile error stating that the MaterialTheme.colorScheme can only be used in a composable context.

Reply:
ahhh yeah fair enough, makes sense, possibly a shortcoming of this wrapped data class approach to discuss... 🤔